autoload/suda.vim

```diff
@@ -199,6 +198,41 @@ function! suda#FileWriteCmd() abort
   endtry
 endfunction
 
+function! suda#BufEnter() abort
+  let bufname = expand('<afile>')
+  let bufnr = bufnr('%')
+  if !empty(&buftype) || empty(bufname) || match(bufname, '^[a-z]\+://*') isnot# -1
+    " Non file buffer
+    return
+  elseif filereadable(bufname) && filewritable(bufname)
+    " File is readable and writeable
+    return
+  elseif empty(getftype(bufname))
+    " if file doesn't exist, we search for a all directories up it's path to
+    " see if each one of them is writable, if not, we `return`
+    let parent = fnamemodify(bufname, ':p:h')
+    while 1
+      if filewritable(parent) is# 2
+        " reached a writeable directory - no need to switch to suda://
+        return
+      elseif !filereadable(parent) && isdirectory(parent)
+        " reached a readable parent directory which is not writeable so it
+        " means we finally can be sure we need to replace our buffer with a
+        " suda:// .
+        break
+      endif
+      " we keep on the iteration by going up the tree
+      let parent = fnamemodify(parent, ':h')
+    endwhile
```

Comment:
Because sometimes, you may find yourself opening a file under a writeable directory e.g.  `~/all/of/these/directories/do/not/exist.txt` and then you would like to edit that buffer (of course without `suda://`) and when it is ready for save, `:!mkdir -p %:p:h` and then `:w`.

Reply:
Even without suda.vim, `:e ~/all/of/.these/directories/do/not/exist.txt` does not work anyway so suda.vim should NOT consider that case I think.

Reply:
> `:e ~/all/of/these/directories/do/not/exist.txt` does not work anyway

What do you mean by 'does not work anyway?

Reply:
You've already explained it by yourself. Simple `:w` won't work.

I believe that procedure without automatic `suda://` prefix is more simple

### Unwritable without automatic `suda://` prefix

```
:e /all/of/these/directories/do/not/exist.txt
:!sudo mkdir -p %:p:h
password
:w suda://%
```

### Unwritable with automatic `suda://` prefix

```
:e /all/of/these/directories/do/not/exist.txt
:!sudo mkdir -p <C-r>=fnamemodify(matchstr(expand('%'), '^suda://\zs.*'), ':p:h')
password
:w
```

Reply:
My bad. I think I got the reason for this.

Well, could you use the following instead in case? It made sure that the loop stop on root.

```vim
    " if file doesn't exist, we search for a all directories up it's path to
    " see if each one of them is writable, if not, we `return`
    let parent = fnamemodify(bufname, ':p')
    while parent !=# fnamemodify(parent, ':h')  " stop loop on root
      let parent = fnamemodify(parent, ':h')
      if filewritable(parent) is# 2
        " reached a writeable directory - no need to switch to suda://
        return
      elseif !filereadable(parent) && isdirectory(parent)
        " reached a readable parent directory which is not writeable so it
        " means we finally can be sure we need to replace our buffer with a
        " suda:// .
        break
      endif
    endwhile
```